text/0000-wire-reform.md

```diff
@@ -0,0 +1,146 @@
+
+# Goals
+
+The primary goal of this RFC is to decouple the wire service from the LWC and from the LWC Wire Decorator implementation.
+
+As a secondary goal is to embrace reactivity for the configuration payload and the wired method in LWC.
+
+A third goal is to support the provision of wire adapters via wire service on any object, whether it is LWC component or not.
+
+# No-goals
+
+* This reform does not change the wire decorator syntax.
+* This reform does not change the wire adapter API for lightning platform (we should be able to keep that intact).
+
+# Proposal
+
+This reform is focused on the refactor of the wire decorator code, and the wire-service code. As part of the separation process, there are certain responsibilities that must be well defined:
+
+## Responsibilities of the wire decorator
+
+* To install a prototype descriptor to handle the wired field value from the `vm` of the component.
+* To create an instance of an adapter and link it to the host.
+* To signal to the adapter instance when the component is connected or disconnected.
+* To signal to the adapter instance when the config have changed by providing the new config object.
+* To extract the config value from the host object by relying on the compiler's wire metadata.
+* To signal to the adapter instance when a context is available by providing the new context value. 
+
+## Responsibilities of the wire service
+
+* To define the wire adapter protocol.
+* To provide a referral implementation of the wire adapter protocol.
+* To provide an optional abstraction for existing "legacy" wire adapter factories.
+
+## Implementation Details
+
+* `@lwc/engine` does not know about `@lwc/wire-service` and vice versa.
+* `@lwc/engine` will install a descriptor on the prototype for every wired field during the decorators registration routine.
+* `@wire` decorator will create an instance of the provided `WireAdapter` during the component initialization routine by providing the data callback, and the context callback as arguments. 
+* `@wire` decorator will invoke the adapter's `connect` and `disconnect` based on the internal hooks used per component.
+* `@wire` decorator will create a mutation tracking phase to track any access executed during the computation of the config before calling the `update` routine to be able to detect mutations on those values and issue another update on the adapter instance.
+* `@wire` decorator will implement the logic to provide contextual information when requested by the wire adapter.
+* `@lwc/compiler` will provide a config function per `@wire()` declaration to produce a new config object when invoked with the `component` as the first argument. The `@wire` adapter can rely on that config function to produce a new config object at will.
+* `@lwc/wire-service` becomes lightning platform specific for most part (`register()` method), since anyone can implement the wire adapter protocol.
```

Comment:
Since this proposal drops the wire adapter registration, is the `@lwc/wire-service` package will only be used for legacy wires. Once all the wire adapters have been updated we should get rid of this package.

Reply:
YES! I believe that package will become just a referral implementation, an optional base implementation, and a bunch of types.